### PR TITLE
CC-1281: Add SingleLineOutputTestCase and refactor test cases

### PR DESCRIPTION
**Pull request by ryan-gang**
This pull request adds a new test case called SingleLineOutputTestCase and refactors the existing test cases to use this new test case instead of the RegexTestCase. The SingleLineOutputTestCase verifies that a prompt exists, sends a command, and matches the output against a string. This change improves the readability and maintainability of the test cases.

### CODE REVIEW

**Comment by ryan-gang** [comment]
Deprecated `regex_test_case`, and integrated `single_line_output_test_case` in all the stages.

File: internal/test_cases/single_line_output_test_case.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/codecrafters-io/tester-utils/logger"
+)
+
+// SingleLineOutputTestCase verifies a prompt exists, sends a command and matches the output against a string.
+type SingleLineOutputTestCase struct {
+	// The command to execute (the command's output will be matched against ExpectedPattern)
+	Command string
+
+	// ExpectedPattern is the regex that is evaluated against the command's output.
+	// Add \r\n at the end of the pattern if you're expecting a newline.
```
Comment (rohitpaulk) [inline internal/test_cases/single_line_output_test_case.go:20]: Don't think that this ("Add \r\n at the end") needs to be a requirement anymore. If we're expecting a single line, wouldn't it always have a `\r\n` at the end?

File: internal/test_cases/single_line_output_test_case.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+			return fmt.Errorf("Expected first line of output to end with '\\n' (newline), got %q", string(cleanOutput(output)))
+		} else if errors.Is(err, shell_executable.ErrProgramExited) {
+			return fmt.Errorf("Expected shell to be a long-running process, but it exited")
+		}
+		return err
+	}
+
+	// We will explicitly support the case where user uses CR/LF at the end
+	// and we receive CR/CR/LF
+	if len(output) > 2 && bytes.Equal(output[len(output)-3:], []byte{'\r', '\r', '\n'}) {
```
Comment (rohitpaulk) [inline internal/test_cases/single_line_output_test_case.go:64]: Seems very specific to check for just these two cases, is there a more "general" way to handle this? 

For example, maybe something like `stripLineEnding` that takes bytes, and then:

- Collapses all consecutive `\r`s into a single `\r` (it doesn't matter how many there are really, they're equivalent)
- Removes `\r\n` from the end of the string

Comment (rohitpaulk) [reply internal/test_cases/single_line_output_test_case.go:64]: (I say "remove \r\n" assuming that we won't have `ExpectedPattern` include it, separate comment above)

File: internal/test_cases/single_line_output_test_case.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func reversePTYTransformation(buf []byte) []byte {
+	// PTY converts all bare LFs to CRLFs, so we just reverse that here
+	// Bare CRs are left as is
+	buf = bytes.ReplaceAll(buf, []byte{'\r', '\n'}, []byte{'\n'})
+	return buf
+}
+
+func cleanOutput(buf []byte) []byte {
```
Comment (rohitpaulk) [inline internal/test_cases/single_line_output_test_case.go:96]:
```suggestion
func formatOutputForLogging(buf []byte) []byte {
```

(Might not be immediately clear whether the output is being "cleaned" for pattern matching or for logging)

File: internal/test_cases/single_line_output_test_case.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		return fmt.Errorf("Expected first line of output to %s, got %q", t.ExpectedPatternExplanation, string(cleanOutput(output)))
+	}
+
+	logger.Successf("✓ %s", t.SuccessMessage)
+
+	return nil
+}
+
+func stripSpaceRight(buf []byte) []byte {
```
Comment (rohitpaulk) [inline internal/test_cases/single_line_output_test_case.go:83]:
```suggestion
func stripLineEnding(buf []byte) []byte {
```

("space" isn't an accurate term here)

**Review by rohitpaulk — APPROVED** [review]

**Comment by rohitpaulk** [comment]
> Let me know if this works: [16adc82](https://github.com/codecrafters-io/shell-tester/pull/24/commits/16adc829c135ac6b1fd0e4818a0b2e9479fcb941)

@ryan-gang yep, we need this. We can just add `$` to the regex string passed in though. We're adding `^` already at the start, this is similar. If an interface accepts a regex, I think it's reasonable to expect that without `^`/`$` it'd just match a substring

**Review by rohitpaulk — APPROVED** [review]

**Comment by ryan-gang** [comment]
Fine. 
Will update all the regexes instead, to pass `$`.

**Pull request merged**